pkg/cid/cid.go

```diff
@@ -85,6 +85,12 @@ func New(stub ChaincodeStubInterface) (ClientIdentity, error) {
 
 // GetID returns a unique ID associated with the invoking identity.
 func (c *clientIdentityImpl) GetID() (string, error) {
+	// When IdeMix, c.cert is nil for x509 type
+	// Here will return "", as there is no x509 type cert for generate id value with logic below.
+	cert, _ := c.GetX509Certificate()
+	if (cert == nil){
+		return "",nil
```

Comment:
Is returning an empty identity the right thing to do here? Certainly we don't want to panic, but, perhaps a typed error indicating that there is no identity available would be the superior option?

Reply:
thanks jyellick, here is my point.
1. I think we got agreement on avoid panic when cert is null when submitter msp in idemix type.
2. I didn't know further plan for CID interface. To be specific, current CID is generate by cert info. 
2.1 Q: Will we have some way to get the same info from idemix msp?
2.2 Q: Will we change CID generate logic to make it supports both x509 and idemix? ex: depends on role and ou?
You can share me the further plan and I would like to implements in this PR.
3. With unknown in above, for string value part, as cert is empty, so I just make it as an empty string. For error part, raising up a new error "This tx is submitted by an idemix msp" is a good option for chaincode logic?  I don't think chaincode should do specific for idemix type msp with an error type, `if(err = idemix) then `. But, `if(idemix) then`. So I return nil here for error.

Reply:
1. Definitely! Let's not panic when the identity is supplied via idemix.

2.1: This is a good question, in general, the purpose of idemix is to provide unlinkability of identities.  So, given two different invocations by the same user, the ClientID would be expected to return different values.  I'm unsure whether there's a sensible ClientID value idemix could provide.  @adc @ale-linux ?
2.2: To me, an ID is a something that has a 1-1 relationship with a user.  So, if we were to base ID on role, and necessarily multiple users share the same role, it doesn't quite fit to me.  As I see it, for an idemix ID, we can either take an approach that "idemix is explicitly to anonymize IDs, so GetClientID doesn't make sense", or, we can take an approach that "each idemix invocation effectively forms a new psuedo-client ID which will be unique at each invocation".  To me, GetClientID seems most useful for tracking ownership and associating state across multiple requests, so, I'm more inclined to say that chaincodes relying on ID are likely already not compatible with idemix.

3.  Yes, if we go with an error (which I think is the right thing to do), then we can simply do something like:

```
// ClientIDUnresolvableError indicates that the identity provider cannot map the
// client's identity to a stable ID.  This error is returned by membership service providers
// such as IdentityMixer which provide unlinkability across invocations.
var ClientIDUnresolvableError = errors.New("generating a ClientID is not supported by this identity's membership service provider")

...
    if c.cert == nil {
        return "", ClientIDUnresolvableError
    }
```

@sykesm I know there are pros and cons, but do you think a package level constant like the above would be appropriate here?  Do we have a better approach for exported testable error types?

4.  Yes, in general, if the identity provider is a classical X509 MSP, then if the cert is nil, we'll never get this far in the code. So (today) that error condition is logically equivalent to an Idemix identity, but, it seems easy enough to generalize to "any identity provider which cannot provide client ID".

Reply:
@jyellick , 1st of all, this pr fixed the panic. 2nd to completed fix the panic we need solution for question 2 to 4. 
by summary 2 to 4
`func (c *clientIdentityImpl) GetID() (string, error) `
should return `ClientIDUnresolvableError` as a general fail case, and handle msp type for both x509 and idemix, making a one one mapping for identity and ID string.

I don't think the solution for idemix logic is a show stopper for this PR. 
As I created in https://jira.hyperledger.org/browse/FAB-16948
The major goal is to fix panic.

Reply:
@SamYuan1990 yes, this is definitely a bug, and should definitely be fixed.  We appreciate you opening the bug (and better yet submitting this pull request!).

My concern is that if a chaincode is using the CID to for instance test if an invoker owns an asset, that by returning `""` for all idemix identities with no error, the chaincode could accidentally treat all idemix users as the same underlying user.

I definitely do want to see this PR merged.  If you'd rather simply have this PR address the panic, then please return `"", fmt.Errorf("cannot determine identity")` and we can discuss further exactly how we want to deal with idemix and CIDs in a later PR.

Reply:
Thanks @jyellick , just updated here: https://github.com/hyperledger/fabric-chaincode-go/pull/9/files#diff-b4bca3d85d7311d691afcc297342a077R91

Would you mind me to create another jira ticket for further discussion?

Reply:
Thanks for the update, looks good to me.  Feel free to open a JIRA to track the further enhancements and or discussion.

Reply:
@jyellick , here you are https://jira.hyperledger.org/browse/FAB-16967